File: Linked Lists/list.cc
using namespace std;

#include <iostream>
#include <vector>

template <typename T>
class Node {
 public:
  T data;
  Node* next;

  Node(T d) {
    this->data = d;
    this->next = NULL;
  }
};

template <typename T>
class LinkedList {
 public:
  Node<T>* head;

  LinkedList() { this->head = NULL; }

  Node<T>* getHead() {
    if (!head) {
      return (Node<T>*)NULL;
    }
    return head;
  }

  Node<T>* getTail() {
    if (!head) {
      return (Node<T>*)NULL;
    }
    Node<T>* lastNode = head;
    while (lastNode->next) {
      lastNode = lastNode->next;
    }
    return lastNode;
  }

  void prepend(T d) {
    Node<T>* newNode = new Node<T>(d);
    newNode->next = head;
    head = newNode;
  }

  void append(T d) {
    Node<T>* newNode = new Node<T>(d);

    if (!head) {
      head = newNode;
      return;
    }

    Node<T>* lastNode = head;
    while (lastNode->next) {
      lastNode = lastNode->next;
    }

    lastNode->next = newNode;
  }

  void insertBetween(T data, T before, T after) {
    Node<T>* prevNode = head;
    Node<T>* nextNode = head;

    Node<T>* newNode = new Node<T>(data);

    while (prevNode->data != after) {
      prevNode = prevNode->next;
    }

    while (nextNode->data != before) {
      nextNode = nextNode->next;
    }

    prevNode->next = newNode;
    newNode->next = nextNode;
  }

  void deleteHead() {
    Node<T>* curr = head;
    if (curr) {
      head = curr->next;
      curr = NULL;
      delete curr;
      return;
    }
  }

  void deleteTail() {
    if (!head) {
      return;
    }
    if (!head->next) {
      head = NULL;
      delete head;
      return;
    }

    Node<T>* prevNode = head;

    while (prevNode->next->next) {
      prevNode = prevNode->next;
    }
    prevNode->next = NULL;
    delete prevNode->next;
  }

  void deleteAny(T target) {
    Node<T>* curr = head;

    if (curr && curr->data == target) {
      head = curr->next;
      curr = NULL;
      delete curr;
      return;
    }

    Node<T>* prevNode = NULL;

    while (curr) {
      if (curr->data == target) {
        break;
      }
      prevNode = curr;
      curr = curr->next;
    }

    if (!curr) {
      return;
    }

    prevNode->next = curr->next;
    curr = NULL;
    delete curr;
  }

  int size() {
    int count = 0;
    Node<T>* curr = head;
    while (curr) {
      count++;
      curr = curr->next;
    }
    return count;
  }

  void reverseList() {
    Node<T>* prev = NULL;
    Node<T>* curr = head;
    Node<T>* nxt = NULL;
    while (curr) {
      nxt = curr->next;
      curr->next = prev;
      prev = curr;
      curr = nxt;
    }
    head = prev;
  }

  Node<T>* find(T target) {
    Node<T>* curr = head;
    Node<T>* toReturn = NULL;
    bool found = false;

    while (curr) {
      if (curr->data == target) {
        found = true;
        toReturn = curr;
        break;
      }
      curr = curr->next;
    }

    return (found) ? curr : (Node<T>*)NULL;
  }

  void printList() {
    Node<T>* curr = head;
    while (curr) {
      cout << curr->data << " ";
      curr = curr->next;
    }
    cout << endl;
  }

  void clear() {
    Node<T>* curr = head;
    Node<T>* nxt = NULL;
    while (curr) {
      nxt = curr->next;
      curr = NULL;
      delete curr;
      curr = nxt;
    }
    head = NULL;
  }
};

File: js/LinkedList/doubly/list.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
        this.nodeMap = new Map();
    }

    getHead() {
        if (!this.head) throw Error("List is empty!");
        return this.head;
    }

    getTail() {
        if (!this.head) throw Error("List is empty!");
        return this.tail;
    }

    prepend(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.head.prev = newNode;
            newNode.next = this.head;
            this.head = newNode;
        }
        this.nodeMap.set(data.id, newNode)
        this.length++;
    }

    append(data) {
        const newNode = new Node(data);
        if (!this.head) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.nodeMap.set(data.id, newNode);
        this.length++;
    }

    insertBetween(data, before, after) {
        const newNode = new Node(data);
        let curr = this.head;

        // finding the node with the `after` data
        while (curr && curr.data.id !== after.id) {
            curr = curr.next;
        }

        if (curr) {
            // finding the node with the `before` data
            let nextNode = curr.next;
            while (nextNode && nextNode.data.id !== before.id) {
                curr = nextNode;
                nextNode = nextNode.next;
            }

            // if the `before` node is found, insert the new node
            if (nextNode && nextNode.data.id === before.id) {
                newNode.next = nextNode;
                newNode.prev = curr;
                curr.next = newNode;
                nextNode.prev = newNode;
                this.nodeMap.set(data.id, newNode);
                this.length++;
            } else {
                throw Error("Before node not found");
            }
        } else {
            throw Error("After node not found");
        }
    }

    deleteHead() {
        if (!this.head) {
            throw Error("List is empty!");
        } else {
            this.nodeMap.delete(this.head.data.id);
            this.head = this.head.next;
            if (this.head) {
                this.head.prev = null;
            } else {
                this.tail = null;
            }
            this.length--;
        }
    }

    print() {
        let curr = this.head;
        let result = []
        while (curr) {
            result.push(curr.data.toString());
            curr = curr.next;
        }
        console.log(result.join(" <-> "));
    }
}